src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -198,820 +567,1087 @@ private void checkDataLength(int ... lengths) {
-                ghashAllToS.update(ct, ctOfs, tlen);
-                inOfs += tlen;
-                outOfs += tlen;
-                ctOfs += tlen;
-                i++;
-            }
-            ilen -= count * plen;
-            processed += count * plen;
-        }
-
-        gctrPAndC.doFinal(in, inOfs, ilen, out, outOfs);
-        processed += ilen;
-
-        int lastLen = ilen % AES_BLOCK_SIZE;
-        if (lastLen != 0) {
-            ghashAllToS.update(ct, ctOfs, ilen - lastLen);
-            ghashAllToS.update(
-                    expandToOneBlock(ct, (ctOfs + ilen - lastLen), lastLen));
-        } else {
-            ghashAllToS.update(ct, ctOfs, ilen);
-        }
-    }
-
-    // Process en/decryption all the way to the last block.  It takes both
-    // For input it takes the ibuffer which is wrapped in 'buffer' and 'src'
-    // from doFinal.
-    void doLastBlock(ByteBuffer buffer, ByteBuffer src, ByteBuffer dst)
-        throws IllegalBlockSizeException {
-
-        if (buffer != null && buffer.remaining() > 0) {
-            // en/decrypt on how much buffer there is in AES_BLOCK_SIZE
-            processed += gctrPAndC.update(buffer, dst);
-
-            // Process the remainder in the buffer
-            if (buffer.remaining() > 0) {
-                // Copy the remainder of the buffer into the extra block
-                byte[] block = new byte[AES_BLOCK_SIZE];
-                int over = buffer.remaining();
-                int len = over;  // how much is processed by in the extra block
-                buffer.get(block, 0, over);
-
-                // if src is empty, update the final block and wait for later
-                // to finalize operation
-                if (src.remaining() > 0) {
-                    // Fill out block with what is in data
-                    if (src.remaining() > AES_BLOCK_SIZE - over) {
-                        src.get(block, over, AES_BLOCK_SIZE - over);
-                        len += AES_BLOCK_SIZE - over;
+                    // If src has data, complete the block;
+                    int slen = Math.min(src.remaining(), blockSize - over);
+                    if (slen > 0) {
+                        src.get(block, over, slen);
+                    }
+                    int len = slen + over;
+                    if (len == blockSize) {
+                        resultLen += op.update(block, 0, blockSize, dst);
                     } else {
-                        // If the remaining in buffer + data does not fill a
-                        // block, complete the ghash operation
-                        int l = src.remaining();
-                        src.get(block, over, l);
-                        len += l;
+                        resultLen += op.doFinal(block, 0, len, block,
+                                0);
+                        if (dst != null) {
+                            dst.put(block, 0, len);
+                        }
```

Comment:
Can this be "resultLen += op.doFinal(block, 0, len, dst)"?

Reply:
doFinal doesn't have a (byte[], int, int, ByteBuffer) method.  While that's not a bad idea to have one, it would be a fair bit of code to do it because it's part of the GCM interface and I'd have to write methods for GCTRGHASH, GCTR, and GHASH.  I think that's  too much just for this one code segment that isn't broken.

Reply:
Sure, sounds reasonable.